lightning/src/routing/onion_message.rs

```diff
@@ -0,0 +1,185 @@
+pub fn find_path<L: Deref, GL: Deref>(
+	our_node_pubkey: &PublicKey, destination: &PublicKey, network_graph: &NetworkGraph<GL>, first_hops: Option<&[&ChannelDetails]>, logger: L
+) -> Result<Vec<PublicKey>, Error> where L::Target: Logger, GL::Target: Logger
+{
+	log_trace!(logger, "Searching for an onion message path from origin {} to destination {} and {} first hops {}overriding the network graph",
+		our_node_pubkey, destination, first_hops.map(|hops| hops.len()).unwrap_or(0),
+		if first_hops.is_some() { "" } else { "not " });
+	let graph_lock = network_graph.read_only();
+	let network_channels = graph_lock.channels();
+	let network_nodes = graph_lock.nodes();
+	let our_node_id = NodeId::from_pubkey(our_node_pubkey);
+	let dest_node_id = NodeId::from_pubkey(destination);
+
+	// Add our start and first-hops to `frontier`.
+	let start = NodeId::from_pubkey(&our_node_pubkey);
+	let mut frontier = BinaryHeap::new();
+	frontier.push(PathBuildingHop { cost: 0, node_id: start, parent_node_id: start });
+	if let Some(first_hops) = first_hops {
+		for hop in first_hops {
+			if !hop.counterparty.features.supports_onion_messages() { continue; }
+			let node_id = NodeId::from_pubkey(&hop.counterparty.node_id);
+			frontier.push(PathBuildingHop { cost: 1, node_id, parent_node_id: start });
+		}
+	}
+
+	let mut visited = HashMap::new();
+	while let Some(PathBuildingHop { cost, node_id, parent_node_id }) = frontier.pop() {
+		match visited.entry(node_id) {
+			hash_map::Entry::Occupied(_) => continue,
+			hash_map::Entry::Vacant(e) => e.insert(parent_node_id),
+		};
+		if node_id == dest_node_id {
+			return Ok(reverse_path(visited, our_node_id, dest_node_id, logger)?)
+		}
+		if let Some(node_info) = network_nodes.get(&node_id) {
+			for scid in &node_info.channels {
+				if let Some(chan_info) = network_channels.get(&scid) {
+					if let Some((_, successor)) = chan_info.as_directed_from(&node_id) {
+						// We may push a given successor multiple times, but the heap should sort its best entry
+						// to the top. We do this because there is no way to adjust the priority of an existing
+						// entry in `BinaryHeap`.
+						frontier.push(PathBuildingHop {
+							cost: cost + 1,
+							node_id: *successor,
+							parent_node_id: node_id,
+						});
+					}
+				}
+			}
+		}
+	}
+
+	Err(Error::PathNotFound)
+}
+
+#[derive(Debug, PartialEq)]
+/// Errored running [`find_path`].
+pub enum Error {
+	/// We failed to find a path to the destination.
+	PathNotFound,
+	/// We failed to convert this node id into a [`PublicKey`].
+	InvalidNodeId(secp256k1::Error),
+}
+
+#[derive(Eq, PartialEq)]
+struct PathBuildingHop {
+	cost: u64,
+	node_id: NodeId,
+	parent_node_id: NodeId,
+}
+
+impl PartialOrd for PathBuildingHop {
+	fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+		// We need a min-heap, whereas `BinaryHeap`s are a max-heap, so compare the costs in reverse.
+		other.cost.partial_cmp(&self.cost)
```

Comment:
Assuming the cost is equal, can someone grind lower node-id to attract more message flows? After staring at rust documentation for 5 minutes, I'm not sure.

Reply:
Added a TODO for this

Reply:
So you confirm it's possible by looking at the rust doc?

Reply:
Perhaps fine to keep a todo anyway because:
- the way nodes sorted before insertion into heap might play a role then
- heap implementation may change